### ISSUE TITLE
SaltelliSensitivityAlgorithm with LowDiscrepancyExperiment produces wrong results

### ISSUE THREAD

**Issue by mbaudin47** [openturns/openturns#1000]
I want to estimate Sobol' indices of the Ishigami function with a low discrepancy sequence.

The following script compares three results
* the exact numerical values,
* the indices produced by SobolIndicesExperiment,
* the indices produced by LowDiscrepancyExperiment.
It seems to me that LowDiscrepancyExperiment produces wrong Sobol' indices.

The following script :
```
import openturns as ot

formula = ['sin(X1*pi_)+7.*sin(X2*pi_)^2+0.1*(X3*pi_)^4*sin(X1*pi_)']

model = ot.SymbolicFunction(['X1', 'X2', 'X3'], formula)

# distributions de X1, X2, X3
dist_X1 = ot.Uniform(-1., 1.)
distribution = ot.ComposedDistribution([dist_X1]*3)

exact_S = [0.3139, 0.4424, 0]
exact_ST = [0.5576,0.4424,0.2437]
print("Exact S = %s" % (str(exact_S)))
print("Exact ST = %s" % (str(exact_ST)))

# Size of simulation
size = 100000

# Test with the various implementation methods

# Use of Generate method to build input/output designs
inputDesign = ot.SobolIndicesExperiment(distribution, size, True).generate()
outputDesign = model(inputDesign)
sensitivity_algorithm = ot.SaltelliSensitivityAlgorithm(inputDesign, outputDesign, size)

print("SobolIndicesExperiment S = %s" % (str(sensitivity_algorithm.getFirstOrderIndices())))
print("SobolIndicesExperiment ST = %s" % (str(sensitivity_algorithm.getTotalOrderIndices())))

# with experiment
sequence = ot.SobolSequence(distribution.getDimension())
experiment = ot.LowDiscrepancyExperiment(sequence, distribution, size)
sensitivity_algorithm = ot.SaltelliSensitivityAlgorithm(experiment, model)
print("SobolSequence S = %s" % (str(sensitivity_algorithm.getFirstOrderIndices())))
print("SobolSequence ST = %s" % (str(sensitivity_algorithm.getTotalOrderIndices())))
```
produces
```
Exact S = [0.3139, 0.4424, 0]
Exact ST = [0.5576, 0.4424, 0.2437]
SobolIndicesExperiment S = [0.3082,0.4367,-0.004091]
SobolIndicesExperiment ST = [0.5628,0.4443,0.2423]
SobolSequence S = [0.3295,0.75558,0.555788]
SobolSequence ST = [0.01268,0.4322,0.2379]
```
It appears that both the first and the total order indices from the SobolSequence are wrong. Something is even more weird : if you exchange the 3d first order indice with the 1st total order indice, you get a very good result ! Is there any chance that the indices were mixed in the code ?

**Comment by mbaudin47** [comment]
This is with OT 1.12.

**Comment by regislebrun** [comment]
Things are more involved:
  * Even if you swap the 3rd first order and 1st total order indices, your results are bad as the 2nd first order index is clearly wrong
  * If you change the SobolSequence for the HaltonSequence, you also get wrong results, but this time no value looks like it has been swaped:
```
SobolSequence S = [0.591251,0.996512,0.588086]
SobolSequence ST = [0.0031817,0.408731,5.87176e-07]
```
  * If you use a MonteCarloExperiment instead of a LowDiscrepancyExperiment, you get the correct results so the indices are not exchanged as this difference is only seen in the constructor of the SobolIndicesExperiment, which is called by the SobolAlgorithmImplementation constructor, and after that the computations are exactly the same
  * If you set the *randomize* flag to True in your LowDiscrepancyExperiment you get correct results:
```
SobolSequence S = [0.316293,0.442871,0.00235069]
SobolSequence ST = [0.555172,0.440706,0.24244]
```

So the bug is somewhere else. As a matter of fact, I am not convinced at all that one can simply replaces calls to a pseudo-random numbers generator by calls to a low discrepancy sequence and get correct results from a mathematical point of view. It works when only the equirepartition is required, such as the computation of a unique expectation, but when several joint expectations are required I suspect that a kind of nonzero correlation is induced somewhere which leads to wrong results. The use of randomization forces this correlation to be zero (asymptotically)

If you want to use a low discrepancy experiment with the Sobol sequence without bothering with its explicit definition and the manipulation of a flag, you can use the constructor based on a distribution but with the *SobolIndicesExperiment-SamplingMethod* key of the ResourceMap set to "QMC".

All these experiments have been done using OT 1.2 except for the use of the "QMC" key, which has been introduced in OT 1.3

**Comment by mbaudin47** [comment]
Ok : the bug is here :

https://github.com/openturns/openturns/blob/951ed07738b6c22acf9ee439e63463392e8be0ff/lib/src/Uncertainty/Algorithm/Sensitivity/SobolIndicesExperiment.cxx#L113

This confirms the PM we exchanged.

Notice that the code uses : 
<pre>
LowDiscrepancyExperiment sobolExperiment(SobolSequence(dimension), distribution, size)
</pre>
This leads to an inconsistent ways of estimating the indices. 
* This is because the pick-freeze algorithm in SaltelliSensitivityAlgorithm generates two different samples of size 'size'. Because we use a LowDiscrepancy sequence, these two samples are not independent and leads to a wrong estimate of the Sobol' indices.
* The correct way of estimating the indices is to generate a SobolSequence(dimension*2), distribution, size). Then the two samples in the pick-freeze estimator must be those corresponding to the components A = [0,...,dimension-1] and B = [dimension,...,2*dimension-1]. 

This generates an architecture issue : e.g. the Saltelli estimator for the Sobol Sequence is different from the Saltelli estimator for the MonteCarloExperiment, at least in the way it generates the two samples.

**Comment by mbaudin47** [comment]
I experimented with the "twice" trick and this is the result : 

![lowdiscrepancysampling](https://user-images.githubusercontent.com/31351465/52149928-ac456d80-266e-11e9-8d8e-2de8579b93a9.png)

The Python POC is here : 

https://github.com/mbaudin47/pycrash/blob/869a6807e939df9ca1eca5cd4ed3bae7cbdc4969/sobol-bench/saltelli-ishigami-QMC.py#L114

**Comment by regislebrun** [comment]
The nice thing is that it works, the bad thing is that I woke up early to implement it without seeing your code! All in all it is fixed, which is the main point ;-)!
We had this discussion on the mailing list with Pamphile Roy (the 20th of June, 2018), and I already saw the trouble with low discrepancy sequences at this time but missed the point when Pamphile explained the method to me. Better late than never...

**Comment by regislebrun** [comment]
Note that the documentation clearly states that both LHSExperiment and LowDiscrepancyExperiment had to be tweaked in order to get meaningful results (which is not an excuse for implementing things the wrong way...)

**Comment by regislebrun** [comment]
Could you please check <https://github.com/openturns/openturns/pull/1002> and tell me if it fits your needs.

**Comment by mbaudin47** [comment]
Thank you very much for the fix. 
* You always use the "twice" trick, instead of just using it when appropriate : this makes the code much simpler.
* You removed the randomization of the QMC and LHS : this is a good thing to do, since this randomization kills the convergence speed. This fixes an extra bug which was not the initial purpose of the ticket. That's two bugfixes for the price of one !
On my side, I benchmarked all the experiments I could get to work (the experiments with unequal weight cannot be used and generate a correct exception) : I think that this fixes bug. 

For the users who cannot wait OT 1.13, I place as attachment a temporary workaround : 

[saltelli-ishigami-QMC-workaround.zip](https://github.com/openturns/openturns/files/2824776/saltelli-ishigami-QMC-workaround.zip)

Since this fixes the bug, I close the ticket.

**Comment by mbaudin47** [comment]
However, I think that there is another bug, even if I do not have the evidence at the moment. The `SobolSimulationAlgorithm` can make use of the asymptotic distribution of the estimator (based on a gaussian). If this algorithm can take a low discrepancy sequence as input (I do not know if this is the case), then the actual distribution of the estimator will not be the one that is expected : in this case, the algorithm may stop earlier or later than expected. I have to check this.

Actually, the `SobolSimulationAlgorithm` can in theory be used, provided that a randomized QMC is used. But, in order to prevent from killing the convergence of the QMC, only a part of the sequence must be randomized, in the following way. 
* Assume that a size n sample is to be used. 
* This integer must be decomposed into two integers : n = nR x nQMC  where nR is the size of the randomization and nQMC is the size of the quasi-Monte-Carlo sequence.
* This is a trade-off in convergence speed : you get the convergence speed of the QMC sequence at a lower rate 1/nQMC (instead of 1/n), but you get a confidence interval with accuracy governed by 1/nR.
* The typical values should be so that nQMC is large and nR is small, because we do not require to have a high accuracy on the confidence interval, while the estimate itself should by as accurate as possible.

I do not know if the current implementation uses this trade-off, but my guess is that this should be the way it should be managed to get the best of both sides. 

This is an issue for the setRandomize method of the LowDiscrepancyExperiment class : the current interface do not allow for such flexibility. Since this discussion is not directly related to the Sobol' indices algorithm, this ticket must be left closed.

**Comment by regislebrun** [comment]
I already noticed the two points:
+ the use of the asymptotic distribution while using non-iid sequences (LHS, QMC)
+ the need to randomize things in a proper way
For the last point, the fact that simulation algorithms work with a separate outer sampling and a block size makes the solution you propose quite straightforward to implement. The randomization of a low discrepancy sequence should be done on a sample basis and not on a point basis, ie the random shift should be the same for all the points of a given generated sample. Easy to implement, I will do it right now.

**Comment by regislebrun** [comment]
My bad: SobolIndicesAlgorithm is *not* an algorithm: it works with a *fixed* sample size. So it is not possible to implement what I had in mind. I checked the code: if the given sampling method is not a Monte Carlo sampling, you already have a test which emits a warning then switched to a bootstrap estimate of the confidence interval. The fix would be to make SobolIndicesAlgorithm act as an actual algorithm, but it is too much work to be done in 10mn.
I fixed the LowDiscrepancyExperiment class to use a global random translation instead of a pointwise random translation, it is already usefull for all the other simulation algorithms.
